src/ripple/app/misc/impl/TxQ.cpp

```diff
@@ -1112,6 +1116,16 @@ TxQ::apply(Application& app, OpenView& view,
         (void)created;
         assert(created);
     }
+    // If the transaction goes into the queue, future attempts need
+    // to be done:
+    // * without RETRY because this will look successful to the caller
+    // * without PREFER_QUEUE because we're already in the queue, so duh
+    // * with NO_CHECK_SIGN because the signature has been checked
+    //   at least once, so don't do it again.
+    // These changes _may_ cause an extra `preflight`, but since signature
+    // checking is the most expensive part of `preflight`, the cost should
+    // be minimal.
+    flags = (flags & ~tapRETRY & ~tapPREFER_QUEUE) | tapNO_CHECK_SIGN;
```

Comment:
I'd prefer we structure this code thusly:

```// If the transaction goes into the queue, future attempts need to adjust the
// flags. This _may_ cause an extra `preflight`, but since signature checking is
// the most expensive part of `preflight`, the cost should be minimal.

// Don't retry, because this will look successful to the caller
flags &= ~tapRETRY;

// Don't queue because we're already in the queue
flags &= ~tapPREFER_QUEUE;

// Don't re-check the signature; once is enough.
flags |= tapNO_CHECK_SIGN;
```

Speaking of which what if an intervening transaction changes the signature requirements for this account? The simplest example is a previous `SetRegularKey` but it can get arbitrarily complex: imagine, for example:

This is a multi-signed tx against account A. Account B is a multisigner for A and a singer for this tx. B uses a regular key, C. B's signature on this transaction is via C.

The transaction is now queued. Before it can be applied again, a new transaction comes in changing B's regular key from C to D. It is successful. The signature of the original transaction is (potentially) no longer valid, yet we won't check again.

Oops! :fearful: